File: core/src/main/python/mmlspark/stages/UDFTransformer.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.

import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from mmlspark.stages._UDFTransformer import _UDFTransformer
from pyspark import keyword_only

from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel

from pyspark.ml.common import inherit_doc
from pyspark.sql.functions import UserDefinedFunction
from pyspark.ml.common import inherit_doc
from mmlspark.core.schema.Utils import *

@inherit_doc
class UDFTransformer(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaTransformer):
    """
    Args:

        inputCol (str): The name of the input column (default: )
        outputCol (str): The name of the output column
        udf (object): User Defined Python Function to be applied to the DF input col
        udfScala (object): User Defined Function to be applied to the DF input col
    """

    @keyword_only
    def __init__(self, inputCol=None, inputCols=None, outputCol=None, udf=None):
        super(UDFTransformer, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.stages.UDFTransformer")
        self.inputCol = Param(self, "inputCol", "inputCol: The name of the input column (default: )")
        self.inputCols = Param(self, "inputCols", "inputCols: The names of the input columns (default: )")
        self.outputCol = Param(self, "outputCol", "outputCol: The name of the output column")
        self.udf = Param(self, "udf", "udf: User Defined Python Function to be applied to the DF input col")
        if udf != None:
            self.setUDF(udf)
            self._udf = udf
        if inputCol != None:
            self.setInputCol(inputCol)
        if inputCols != None:
            self.setInputCols(inputCols)
        if outputCol != None:
            self.setOutputCol(outputCol)

    def setInputCol(self, value):
        """

        Args:

            inputCol (str): The name of the input column (default: )

        """
        self._set(inputCol=value)
        return self


    def getInputCol(self):
        """

        Returns:

            str: The name of the input column (default: )
        """
        return self.getOrDefault(self.inputCol)


    def setInputCols(self, value):
        """

        Args:

            inputCols (list): The names of the input columns (default: )

        """
        self._set(inputCols=value)
        return self

    def getInputCols(self):
        """

        Returns:

            str: The name of the input column (default: )
        """
        return self.getOrDefault(self.inputCols)

    def setOutputCol(self, value):
        """

        Args:

            outputCol (str): The name of the output column

        """
        self._set(outputCol=value)
        return self


    def getOutputCol(self):
        """

        Returns:

            str: The name of the output column
        """
        return self.getOrDefault(self.outputCol)

    def setUDF(self, udf):
        name = getattr(udf,"_name",  getattr(udf, "__name__", None))
        name = name if name else udf.__class__.__name__
        userDefinedFunction = UserDefinedFunction(udf.func, returnType = udf.returnType, name = name)
        self._java_obj = self._java_obj.setUDF(userDefinedFunction._judf)
        self._udf = udf
        return self

    def getUDF(self):
        return self._udf

    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.stages.UDFTransformer"

    @staticmethod
    def _from_java(java_stage):
        module_name=_UDFTransformer.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".UDFTransformer"
        return from_java(java_stage, module_name)

File: core/src/main/scala/com/microsoft/ml/spark/stages/UDFTransformer.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.ml.spark.stages

import com.microsoft.ml.spark.codegen.Wrappable
import com.microsoft.ml.spark.core.contracts.{HasInputCol, HasInputCols, HasOutputCol}
import com.microsoft.ml.spark.core.serialize.ComplexParam
import com.microsoft.ml.spark.logging.BasicLogging
import org.apache.spark.injections.UDFUtils
import org.apache.spark.ml.{ComplexParamsReadable, ComplexParamsWritable, Transformer}
import org.apache.spark.ml.param.{ParamMap, UDFParam, UDPyFParam}
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.sql.execution.python.UserDefinedPythonFunction
import org.apache.spark.sql.expressions.UserDefinedFunction
import org.apache.spark.sql.types.{DataType, StringType, StructField, StructType}
import org.apache.spark.sql.{Column, DataFrame, Dataset}
import org.apache.spark.sql.functions.col

object UDFTransformer extends ComplexParamsReadable[UDFTransformer]

/** <code>UDFTransformer</code> takes as input input column, output column, and a UserDefinedFunction
  * returns a dataframe comprised of the original columns with the output column as the result of the
  * udf applied to the input column
  */
class UDFTransformer(val uid: String) extends Transformer with Wrappable with ComplexParamsWritable
  with HasInputCol with HasInputCols with HasOutputCol with BasicLogging {
  logClass()

  def this() = this(Identifiable.randomUID("UDFTransformer"))

  override protected lazy val pyInternalWrapper = true

  val udfScalaKey = "udfScala"
  val udfPythonKey = "udf"
  val invalidColumnSetError = s"If using multiple columns, only use set inputCols"

  setDefault(outputCol -> (this.uid + "_output"))

  override def setInputCol(value: String): this.type = {
    require(!isSet(inputCols), invalidColumnSetError)
    set(inputCol, value)
  }

  override def setInputCols(value: Array[String]): this.type = {
    require(!isSet(inputCol), invalidColumnSetError)
    set(inputCols, value)
  }

  val udfScala = new UDFParam(this, udfScalaKey, "User Defined Function to be applied to the DF input col")
  val udfPython = new UDPyFParam(this, udfPythonKey,
    "User Defined Python Function to be applied to the DF input col")
  val udfParams: Seq[ComplexParam[_]] = Seq(udfScala, udfPython)

  /** @group getParam */
  def getUDF: UserDefinedFunction = $(udfScala)

  /** @group getParam */
  def getUDPyF: UserDefinedPythonFunction = $(udfPython)

  /** @group setParam */
  def setUDF(value: UserDefinedFunction): this.type = {
    udfParams.filter(isSet).foreach(clear)
    set(udfScalaKey, value)
  }

  /** @group setParam */
  def setUDF(value: UserDefinedPythonFunction): this.type = {
    udfParams.filter(isSet).foreach(clear)
    set(udfPythonKey, value)
  }

  def applyUDF(col: Column): Column =  {
    if (isSet(udfScala)) getUDF(col)
    else getUDPyF(col)
  }

  def applyUDFOnCols(cols: Column*): Column =  {
    if (isSet(udfScala)) getUDF(cols: _*)
    else getUDPyF(cols: _*)
  }

  def getDataType: DataType =  {
    if (isSet(udfScala)) UDFUtils.unpackUdf(getUDF)._2
    else getUDPyF.dataType
  }

  /** @param dataset - The input dataset, to be transformed
    * @return The DataFrame that results from applying the udf to the inputted dataset
    */
  override def transform(dataset: Dataset[_]): DataFrame = {
    logTransform[DataFrame]({
      transformSchema(dataset.schema, logging = true)
      if (isSet(inputCol)) {
        dataset.withColumn(getOutputCol, applyUDF(dataset.col(getInputCol)))
      } else {
        dataset.withColumn(getOutputCol, applyUDFOnCols(getInputCols.map(col): _*))
      }
    })
  }

  def validateAndTransformSchema(schema: StructType): StructType = {
    if (isSet(inputCol)) schema(getInputCol) else schema(Set(getInputCols: _*))
    schema.add(StructField(getOutputCol, getDataType))
  }

  def transformSchema(schema: StructType): StructType = validateAndTransformSchema(schema)

  def copy(extra: ParamMap): UDFTransformer = defaultCopy(extra)

}

